packages/riverpod/lib/src/framework/base_provider.dart

```diff
@@ -608,6 +657,18 @@ class ProviderElement<Created, Listened> implements ProviderReference {
     }
   }
 
+  @override
+  void setState(Listened newState) {
+    if (!_mounted) {
+      throw StateError('Cannot call setState after a provider was dispose');
+    }
+    assert(_debugIsFlushing == false,
+        'Cannot call .setState(newState) while building/onDispose on $_provider');
+    state._previousValue = state._exposedValue;
+    state.exposedValue = newState;
+    markDidChange();
```

Comment:
That won't work with StateNotifiers/ChangeNotifiers.

Doing `setState(StateNotifier())` will not cause it to be listened

Reply:
Do we even want those providers to have a `ref.setState` anyway?
 
 Maybe that functionality should be excluded for these two. If so, there's a bit of extra work to preserve the "compile safety"

Reply:
In my first implementation i created an `AdvancedProviderReference extends ProviderReference {}` including the new features.
`ProviderElement` would extend `AdvancedProviderReference` and `Provider` was the only class to use it.
This would make `Provider` the only one to know anything about the new features. What do you think about making **only** `Provider` more powerful?

Reply:
>  What do you think about making only Provider more powerful?

FutureProvider/StreamProvider definitely wants to benefit from these features though

Reply:
>  What do you think about making only Provider more powerful?

FutureProvider/StreamProvider definitely wants to benefit from these features though

Reply:
So the last few hour I tried to implement this compile safe. The best solution i found is to wrap ProviderReference.
It will allow to hide `.setState` from `StateNotifier` but will still allow to expose `.currentState`

What do you think about wrapping `ProviderReference`?

Here is just a basic code explaining what is happening, basically
```dart
typedef Create<T, Listened, Ref extends WrapRefBase<Listened>> = T
    Function(Ref ref);

class WrapRefBase<Listened> {
 //// Expose mounted, currentState, watch, read
 Listened currentState => ref.currentState;
}

class WrapRefDefault<Listened> extends WrapRefBase<Listened> {
 final ProviderReference ref;
}

class WrapRefAdvanced<Listened> extends WrapRefBase<Listened> {
 final ProviderReference ref;
 //// Expose mounted, currentState, watch, read, setState ...
 void setState(new) => ref.setState(new);
}


class Provider<T> extends AlwaysAliveProviderBase<T, T> {
  /// {@macro riverpod.provider}
  Provider(
    CreateActions<T, T, WrapRefAdvanced<T>> create, {
    String? name,
  }) : super((ref) => create(WrapRefAdvanced<T>(ref)), name);


/// {@macro riverpod.statenotifierprovider}
@sealed
class StateNotifierProvider<T extends StateNotifier<Object?>>
    extends Provider<T> {
  /// {@macro riverpod.statenotifierprovider}
  StateNotifierProvider(
    Create<T, T, WrapRefDefault<T>> create, {
    String? name,
  }) : super((wrap) {
          final controller = create(wrap.ref);
          wrap.onDispose(controller.dispose);
          return controller;
        }, name: name);
}

```

Reply:
That's not good. It's breaking and verbose. We'd need to write `wrap.ref.watch`

Can't we subclass ProviderReference?

Reply:
Have a look at the latest code. I think I got it :)